include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,516 @@
+// the 1D view of things, we really only need
+//
+//    L(x ⊕ y) = L(x) ⊕ L(y)  (1D linearity rule),
+//
+// which is part of the definition of L being a linear function.  The new 1D
+// linearity rule plus stacking/unstacking is equivalent to the earlier
+// N-dimensional linearity rule.
+//
+// That's all we need in order to define linear layouts mathematically!
+//
+// # Comaprison to Nvidia CuTe
+//
+// (Note, I'm not an expert on CuTe; this is my best understanding.)
+//
+// CuTe is a programmatic layout system that's part of Nvidia CUTLASS; see
+// https://github.com/NVIDIA/cutlass/blob/629f465/media/docs/cute/00_quickstart.md
+//
+// LLs and CuTe solve similar problems.  Before CuTe, CUTLASS v2 had many
+// handcrafted layouts, "RowMajor", "VoltaTensorOpMultiplicandCongruous", etc,
+// see https://www.youtube.com/watch?v=QLdUML5MCfE&t=574s.  Each of these was a
+// special case.  CUTLASS v3 introduced CuTe layouts, which are programmable and
+// subsume all of these special cases.  The CUTLASS folks say this simplified
+// CUTLASS, in the same way that we hope LLs will simplify Triton.
+//
+// Like CuTe layouts, LLs are also programmable and composible.  But there are
+// also some differences.
+//
+//  - Dimensions in LLs are named; CuTe dimensions are numbered.
+//  - CuTe layouts can be nested; LLs cannot be.  (Nesting doesn't give CuTe
+//    layouts additional power; any nested layout can be flattened.)
+//  - CuTe layouts support non-power-of-two shapes; LLs do not.  In particular
+//    this means that LLs cannot represent padded layouts.
+//  - In CuTe, swizzling is a separate step applied after specifying a layout.
+//    In LLs, swizzling is part of the layout itself.
+//  - The structure of LLs allows us to programmatically search for layouts that
+//    satisfy certain requirements, for example a shared layout that doesn't
+//    have bank conflicts when read into a particular register layout.  CuTe
+//    expects a human to choose the layout using their brain.
+//  - CuTe emits code that is in the critical path of your CPU and GPU programs,
+//    therefore it needs to be fast.  It uses C++ template magic to specialize
+//    on known-sized dimensions, and so on.  LLs themselves do not need to be
+//    fast; only the emitted `apply` code is on the critical path.
+//  - CuTe requires a CUDA compiler such as nvcc; LLs do not.
+//
+class LinearLayout {
+private:
+  // bases[inDim][outDim][i] = L(0, ..., inDim=2^i, ..., 0)[outDim].  All other
+  // values of L are computed by xor'ing bases together, using the linearity
+  // rule.  In addition:
+  //
+  // - Each inDim has the same set of outDims, in the same order.
+  // - The order of dims is minor-to-major, although this only affects reshape.
+  // - For each inDim, all bases[inDim][<any>].size() are equal.
+  llvm::MapVector<StringAttr /*inDim*/,
+                  llvm::MapVector<StringAttr /*outDim*/, std::vector<int32_t>>>
```

Comment:
I'm of two minds about this.  On the one hand, doing bitwise operations on int32_t is yucky.  On the other hand, the interface becomes more complicated, because we receive a variety of types, not all of which I think should be uint32_t.  So now every interface method needs to choose its types.

In practice it shouldn't matter, so I think I'll leave it as-is on the basis of keeping the API simpler.